# 1.3 Formulating Abstractions with Higher-Order Procedures

## 1.3.1 Procedures as Arguments

## Example 9: Summation

Consider the following three procedures.  
The first computes the $sum$ of the integers from $a$ through $b$:

The second computes the $sum$ of the $cubes$ of the integers in the given range:

The third computes the $sum$ of a sequence of $terms$ in the series  

$$\frac{1}{1 \centerdot 3} + \frac{1}{5 \centerdot 7} + \frac{1}{9 \centerdot 11} + \ldots$$

which converges to $\pi$/8 (very slowly):

Indeed, mathematicians long ago identified the abstraction of $summation$ of a series and invented "sigma notation", for example,to express this concept.

$$\sum_{n=a}^bf(n) = f(a) + \ldots+ f(b)$$

With the aid of the procedures sum as follows,we can express the $mathematical$ $concepts$ of $summation$:

In [1]:
cat 1.3/Example_9/sum_by_recursion.scm

(define (sum term a next b)
  (if (> a b) 0
      (+ (term a)
         (sum term (next a) next b))))


$\bullet$ We can redefine the process $sum$-$integers$ to compute the $sum$ of the integers from $a$ through $b$:

In [3]:
cat 1.3/Example_9/sum_integers.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


### Running Instance:

$\bullet$ Redefine the process $sum$-$cubes$ to compute the sum of the cubes of the integers from a through b:

In [4]:
cat 1.3/Example_9/sum_cubes.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (cube x) (* x x x))
(define (inc n) (+ n 1))

(define (sum-cubes a b)
  (sum cube a inc b))


### Running Instance:

$\bullet$ We can also redefine $pi$-$sum$ in the same way:

In [5]:
cat 1.3/Example_9/pi_sum.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (pi-sum a b)
  (define (pi-term x)
    (/ 1.0 (* x (+ x 2))))
  (define (pi-next x) (+ x 4))
  (sum pi-term a pi-next b))


### Running Instance:

Once we have $sum$, we can use it as a building block in formulating further concepts. For instance, the $definite$ $integral$ of a function $f$ between the limits $a$ and $b$ can be approximated numerically using the formula

$$\int_a^bf=\left[f\left(a + \frac{dx}{2}\right) + f\left(a + dx + \frac{dx}{2}\right) + f\left(a + 2dx + \frac{dx}{2}\right) + \centerdot\centerdot\centerdot\right]dx$$

for small values of $dx$. We can express this directly as a procedure:

In [6]:
cat 1.3/Example_9/definite_integral.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (integral f a b dx)
  (define (add-dx x) (+ x dx))
  (* (sum f (+ a (/ dx 2.0)) add-dx b) dx))


For example,we can let $f$ is $cube$ function:

### Running Instance:

#### The exact value of the integral of cube between 0 and 1 is 1/4.

## Exercise 1.29:
Simpson’s Rule is a more accurate method of numerical integration than the method illustrated above. Using Simpson’s Rule, the integral of a function $f$ between $a$ and $b$ is approximated as

$$\frac{h}{3}(y_0 + 4y_1 + 2y_2 + 4y_3 + 2y_4 + \cdots + 2y_{n-2} + 4y_{n-1} + y_n)$$

where $h = (b -a)/n$,for some even integer n, and $y_k = f(a + kh)$.(Increasing n increases the accuracy of the approximation.)Define a procedure that takes as arguments $f$ , $a$, $b$, and $n$ and returns the value of the integral, computed using Simpson’s Rule.Use your procedure to integrate cube between 0 and 1 (with $n = 100$ and $n = 1000$), and compare the results to those of the integral procedure shown above.

### Answer:

In [7]:
cat 1.3/Exercise_1.29/simpson.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (simpson f a b n)
  (define h (/ (- b a) n))
  (define (y k) (f (+ a (* k h))))
  (define (factor k)
    (cond ((or (= k 0) (= k n)) 1)
          ((odd? k) 4)
          (else 2)))
  (define (term k) (* (factor k) (y k)))
  (define (next k) (+ k 1))
  (if (not (even? n))
      (error "n can't be odd number.")
      (* (/ h 3) (sum term (exact->inexact 0) next n))))


### Running Instance:

## Exercise 1.30: 
The $sum$ procedure above generates a $linear$ $recursion$. The procedure can be rewritten so that the sum is performed $iteratively$. Show how to do this by filling in the missing expressions in the following definition:

### Answer:

In [8]:
cat 1.3/Exercise_1.30/sum_by_iteration.scm

(define (sum term a next b)
  (define (iter a result)
    (if (> a b)
        result
        (iter (next a)
              (+ (term a) result))))
  (iter a 0))
		


## Exercise 1.31:
a. The $sum$ procedure is only the simplest of a vast number of similar abstractions that can be captured as higher-order procedures.Write an analogous procedure called $product$ that returns the product of the values of a function at points over a given range. Show how to define $factorial$ in terms of $product$. Also use product to compute approximations to $\pi$ using the formula

$$\frac{\pi}{4} = \frac{2\centerdot4\centerdot4\centerdot6\centerdot6\centerdot8\cdots}{3\centerdot3\centerdot5\centerdot5\centerdot7\centerdot7\cdots}$$

b. If your product procedure generates a recursive process,write one that generates an iterative process. If it generates an iterative process, write one that generates a recursive process.

### Answer:

$\bullet$ The $product$ procedure as follows

In [9]:
cat 1.3/Exercise_1.31/product_by_recursion.scm

(define (product term a next b)
  (if (> a b) 1
      (* (term a)
         (product term (next a) next b))))


$\bullet$ Redefine $factorial$ procedure in terms of $product$ procedure

In [11]:
cat 1.3/Exercise_1.31/factorial_by_product_recursion.scm

(load "1.3/Exercise_1.31/product_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))
(define (factorial b)
  (product identity 1 inc b))


### Running Instance:

$\bullet$ Use $product$ procedure to compute approximations to $\pi$

In [13]:
cat 1.3/Exercise_1.31/pi_by_product_recursion.scm

(load "1.3/Exercise_1.31/product_by_recursion.scm")

(define (product-pi a b)
  (define (pi-next n) (+ n 2))
  (define (pi-term a)
    ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))
  (product pi-term a pi-next b))


### Running Instance:

Since the product procedure above generates a recursive process;we should write one that generates an $iterative$ process

In [14]:
cat 1.3/Exercise_1.31/product_by_iteration.scm

(define (product term a next b)
  (define (iter a result)
    (if (> a b) result
        (iter (next a) (* (term a) result))))
  (iter a 1))


## Exercise 1.32:
a. Show that $sum$ and $product$ (Exercise 1.31) are both special cases of a still more general notion called $accumulate$ that combines a collection of terms, using some general accumulation function:

$accumulate$ takes as arguments the same term and range specifications as $sum$ and $product$, together with a $combiner$ procedure (of two arguments) that specifies how the current term is to be combined with the accumulation of the preceding terms and a null-value that specifies what base value to use when the terms run out. Write $accumulate$ and show how $sum$ and $product$ can both be defined as simple calls to $accumulate$.

b. If your $accumulate$ procedure generates a recursive process, write one that generates an iterative process.If it generates an iterative process, write one that generates a recursive process.

### Answer:

$\bullet$ The $accumulate$ recursive procedure:

In [15]:
cat 1.3/Exercise_1.32/accumulate_by_recursion.scm

(define (accumulate combiner null-value term a next b)
  (if (> a b) null-value
      (combiner (term a)
                (accumulate combiner
                            null-value
                            term
                            (next a)
                            next
                            b))))


$\bullet$ Redefine the $sum$ procedure by way of the $accumulate$ recursive procedure:

In [16]:
cat 1.3/Exercise_1.32/sum_by_accumulate_recursion.scm

(load "1.3/Exercise_1.32/accumulate_by_recursion.scm")

(define (sum term a next b)
  (accumulate + 0 term a next b))


$\bullet$ Redefine the $product$ procedure by way of the $accumulate$ recursive procedure:

In [17]:
cat 1.3/Exercise_1.32/product_by_accumulate_recursion.scm

(load "1.3/Exercise_1.32/accumulate_by_recursion.scm")

(define (product term a next b)
  (accumulate * 1 term a next b))
